provider/aws: aws_route crash when used with aws_vpc_endpoint

### Terraform Version

0.6.15 

### Affected Resource(s)

- aws_route
- aws_vpc_endpoint

### Terraform Configuration Files
```hcl
resource "aws_vpc" "foo" {
	cidr_block = "10.1.0.0/16"
}

resource "aws_internet_gateway" "foo" {
	vpc_id = "${aws_vpc.foo.id}"
}

resource "aws_route_table" "foo" {
	vpc_id = "${aws_vpc.foo.id}"
}

resource "aws_route" "bar" {
	route_table_id = "${aws_route_table.foo.id}"
	destination_cidr_block = "10.3.0.0/16"
	gateway_id = "${aws_internet_gateway.foo.id}"

	# Forcing endpoint to create before route.
	depends_on = ["aws_vpc_endpoint.baz"]
}

resource "aws_vpc_endpoint" "baz" {
  vpc_id          = "${aws_vpc.foo.id}"
  service_name    = "com.amazonaws.us-west-2.s3"
  route_table_ids = ["${aws_route_table.foo.id}"]
}
```

### Panic Output

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x25c366]

goroutine 383 [running]:
panic(0x10b4040, 0xc820014170)
        /Users/phinze/.gimme/versions/go1.6.darwin.amd64/src/runtime/panic.go:464 +0x3e6
github.com/hashicorp/terraform/builtin/providers/aws.findResourceRoute(0xc8204661a8, 0xc8202ae670, 0xc, 0xc820212e01, 0xb, 0x0,
 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_route.go:378 +0x676
github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsRouteCreate(0xc82025b9e0, 0xd72400, 0xc8202c0900, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_route.go:182 +0xa30
github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc8201e3d00, 0xc820196180, 0xc8202af270, 0xd72400, 0xc8202c0900
, 0x57c601, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/helper/schema/resource.go:146 +0x29e
github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc820240030, 0xc820318640, 0xc820196180, 0xc8202af270, 0x1, 0x0
, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/helper/schema/provider.go:162 +0x1ed
github.com/hashicorp/terraform/terraform.(*EvalApply).Eval(0xc820318740, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval_apply.go:67 +0x6e6
github.com/hashicorp/terraform/terraform.EvalRaw(0x300ab10, 0xc820318740, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc8203bbaa0, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0xea
github.com/hashicorp/terraform/terraform.EvalRaw(0x300a3d8, 0xc8203bbaa0, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc820340cc0, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x67
github.com/hashicorp/terraform/terraform.EvalRaw(0x300a400, 0xc820340cc0, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc8203ba280, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0xea
github.com/hashicorp/terraform/terraform.EvalRaw(0x300a3d8, 0xc8203ba280, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.Eval(0x300a3d8, 0xc8203ba280, 0x300a210, 0xc8205f2ea0, 0x0, 0x0, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x6a
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x132b200, 0xc8202da780, 0x0, 0x0)
        /Users/phinze/go/src/github.com/hashicorp/terraform/terraform/graph.go:198 +0x9a9
github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk.func3(0xc82040fcc0, 0xc82035b840, 0xc82040f4c8, 0xc8202da8a0, 0xc82040f
cd0, 0x132b200, 0xc8202da780, 0xc820470900, 0xc820470960)
        /Users/phinze/go/src/github.com/hashicorp/terraform/dag/dag.go:238 +0xeb
created by github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk
        /Users/phinze/go/src/github.com/hashicorp/terraform/dag/dag.go:247 +0x715
exit status 2
FAIL    github.com/hashicorp/terraform/builtin/providers/aws    8.665s
```

### Important Factoids

The tricky thing here is that to trigger the crash AWS has to return the VPC endpoint route before the newly created route in the response - otherwise the loop exits early and the nil pointer dereference never happens.

cc @pshima

--- a/builtin/providers/aws/resource_aws_route.go
+++ b/builtin/providers/aws/resource_aws_route.go
@@ -375,7 +375,7 @@ func findResourceRoute(conn *ec2.EC2, rtbid string, cidr string) (*ec2.Route, er
 	}
 
 	for _, route := range (*resp.RouteTables[0]).Routes {
-		if *route.DestinationCidrBlock == cidr {
+		if route.DestinationCidrBlock != nil && *route.DestinationCidrBlock == cidr {
 			return route, nil
 		}
 	}

--- a/builtin/providers/aws/resource_aws_route_test.go
+++ b/builtin/providers/aws/resource_aws_route_test.go
@@ -158,6 +158,24 @@ func TestAccAWSRoute_noopdiff(t *testing.T) {
 	})
 }
 
+func TestAccAWSRoute_doesNotCrashWithVPCEndpoint(t *testing.T) {
+	var route ec2.Route
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSRouteDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSRouteWithVPCEndpoint,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSRouteExists("aws_route.bar", &route),
+				),
+			},
+		},
+	})
+}
+
 // Acceptance test if mixed inline and external routes are implemented
 /*
 func TestAccAWSRoute_mix(t *testing.T) {
@@ -363,5 +381,34 @@ resource "aws_instance" "nat" {
   ami = "ami-9abea4fb"
   instance_type = "t2.nano"
   subnet_id = "${aws_subnet.test.id}"
+}
+`)
+
+var testAccAWSRouteWithVPCEndpoint = fmt.Sprint(`
+resource "aws_vpc" "foo" {
+  cidr_block = "10.1.0.0/16"
+}
+
+resource "aws_internet_gateway" "foo" {
+  vpc_id = "${aws_vpc.foo.id}"
+}
+
+resource "aws_route_table" "foo" {
+  vpc_id = "${aws_vpc.foo.id}"
+}
+
+resource "aws_route" "bar" {
+  route_table_id         = "${aws_route_table.foo.id}"
+  destination_cidr_block = "10.3.0.0/16"
+  gateway_id             = "${aws_internet_gateway.foo.id}"
+
+  # Forcing endpoint to create before route - without this the crash is a race.
+  depends_on = ["aws_vpc_endpoint.baz"]
+}
+
+resource "aws_vpc_endpoint" "baz" {
+  vpc_id          = "${aws_vpc.foo.id}"
+  service_name    = "com.amazonaws.us-west-2.s3"
+  route_table_ids = ["${aws_route_table.foo.id}"]
 }
 `)

